### PR TITLE
Mech Red R2

### PR DESCRIPTION
**Pull request by soma-p**
Finished the R3 Code for Mech RED! It will be tested using the pads tomorrow in the lab!

### CODE REVIEW

File: src/main/java/frc/robot/systems/GrabberFSM.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+public class GrabberFSM {
+	/* ======================== Constants ======================== */
+	// FSM state definitions
+	public enum FSMState {
+		LOW_IDLE,
+		ASCENDING,
+        CONTROL,
+        DESCENDING,
+        MID_IDLE
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/GrabberFSM.java]: format - some weird spacing issues throughout file, please fix. check tabs vs spaces if it looks right locally but not here?

File: src/main/java/frc/robot/systems/GrabberFSM.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	 * Decide the next state to transition to. This is a function of the inputs
+	 * and the current state of this FSM. This method should not have any side
+	 * effects on outputs. In other words, this method should only read or get
+	 * values to decide what state to go to.
+	 * @param input Global TeleopInput if robot in teleop mode or null if
+	 *        the robot is in autonomous mode.
+	 * @return FSM state for the next iteration
+	 */
+	private FSMState nextState(TeleopInput input){
+        boolean isAscendingButtonPressed = joystick.getRawButton(0);
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/GrabberFSM.java]: magic num for buttons :(

File: src/main/java/frc/robot/systems/GrabberFSM.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        DESCENDING,
+        MID_IDLE
+	}
+
+
+	/* ======================== Private variables ======================== */
+	private FSMState currentState;
+    private Joystick joystick;
+    private CANSparkMax motor;
+    private static final int thresh = 100;
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/GrabberFSM.java]: constant variables are usually all caps

File: src/main/java/frc/robot/systems/GrabberFSM.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                        return FSMState.MID_IDLE;
+                    }else if(!isAscendingButtonPressed && isDescendingButtonPressed){
+                        return FSMState.DESCENDING;
+                    }else if(isAscendingButtonPressed && !isDescendingButtonPressed){
+                        return FSMState.ASCENDING;
+                    } else{
+                        return FSMState.CONTROL;
+                    }
+                }
+				return FSMState.LOW_IDLE;
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/GrabberFSM.java]: according to your FSM you shouldn't ever go back into LOW_IDLE. if input is null, then I think you should treat it as a "no buttons pressed" case

File: src/main/java/frc/robot/systems/GrabberFSM.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+		ASCENDING,
+        CONTROL,
+        DESCENDING,
+        MID_IDLE
+	}
+
+
+	/* ======================== Private variables ======================== */
+	private FSMState currentState;
+    private Joystick joystick;
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/GrabberFSM.java]: you should be using the shared TeleopInput class. take a look at how mech blue sets this up

**Review by tsoim — COMMENTED** [review]
missing integration with Robot.java - please check out how other subsystems have been doing that!

**Review by AG6GR — DISMISSED** [review]

**Review by AG6GR — APPROVED** [review]
Approved. Please remember to edit the commit message if you do squash and merge.

**Pull request merged**